modules/imgcodecs/src/grfmt_png.cpp

```diff
@@ -300,6 +403,379 @@ bool  PngDecoder::readData( Mat& img )
     return result;
 }
 
+bool PngDecoder::nextPage() {
+    if (m_f)
+    {
+        uint id;
+        CHUNK chunkfcTL;
+
+        id = read_chunk(m_f, &chunkfcTL);
+        if (id == id_fcTL && chunkfcTL.size == 38)
+        {
+            // At this point the old frame is done. Let's start a new one.
+            uint w0 = png_get_uint_32(chunkfcTL.p + 12);
+            uint h0 = png_get_uint_32(chunkfcTL.p + 16);
+            uint x0 = png_get_uint_32(chunkfcTL.p + 20);
+            uint y0 = png_get_uint_32(chunkfcTL.p + 24);
+            int delay_num = png_get_uint_16(chunkfcTL.p + 28);
+            int delay_den = png_get_uint_16(chunkfcTL.p + 30);
+            char dop = chunkfcTL.p[32];
+            char bop = chunkfcTL.p[33];
+            printf("**frame props\n-------------------\n");
+            printf("w0 : %d\n", w0);
+            printf("h0 : %d\n", h0);
+            printf("x0 : %d\n", x0);
+            printf("y0 : %d\n", y0);
+            printf("delay_num : %d\n", delay_num);
+            printf("delay_den : %d\n", delay_den);
+            printf("dop : %d\n", dop);
+            printf("bop : %d\n-------------------\n", bop);
+
+            uchar sig[8];
+            if (fread(sig, 1, 8, m_f))
+                return true;
+        }
+    }
+    return false;
+}
+
+void PngDecoder::compose_frame(uchar** rows_dst, uchar** rows_src, uchar bop, uint x, uint y, uint w, uint h)
+{
+    uint  i, j;
+    int u, v, al;
+
+    for (j = 0; j < h; j++)
+    {
+        uchar* sp = rows_src[j];
+        uchar* dp = rows_dst[j + y] + x * 4;
+
+        if (bop == 0)
+            memcpy(dp, sp, w * 4);
+        else
+            for (i = 0; i < w; i++, sp += 4, dp += 4)
+            {
+                if (sp[3] == 255)
+                    memcpy(dp, sp, 4);
+                else
+                    if (sp[3] != 0)
+                    {
+                        if (dp[3] != 0)
+                        {
+                            u = sp[3] * 255;
+                            v = (255 - sp[3]) * dp[3];
+                            al = u + v;
+                            dp[0] = (sp[0] * u + dp[0] * v) / al;
+                            dp[1] = (sp[1] * u + dp[1] * v) / al;
+                            dp[2] = (sp[2] * u + dp[2] * v) / al;
+                            dp[3] = al / 255;
+                        }
+                        else
+                            memcpy(dp, sp, 4);
+                    }
+            }
+    }
+}
+
+uint PngDecoder::read_chunk(FILE* f, CHUNK* pChunk)
+{
+    uchar len[4];
+    pChunk->size = 0;
+    pChunk->p = 0;
+    if (fread(&len, 4, 1, f) == 1)
+    {
+        pChunk->size = png_get_uint_32(len);
+        if (pChunk->size > PNG_USER_CHUNK_MALLOC_MAX)
+            return 0;
+        pChunk->size += 12;
+        pChunk->p = new uchar[pChunk->size];
```

Comment:
Please do not use new as it makes its hard to track memory. Use std::vector<char> if you can.